provider/server.cpp

```diff
@@ -233,24 +239,370 @@ handle_return (
+            }
+        }
+        else
+        {
+            std::ostringstream strm;
+            strm << "set flags for bind socket failed: " << errnoText;
+            SCX_BOOKEND_PRINT (strm.str ().c_str ());
+        }
+    }
+    else
+    {
+        std::ostringstream strm;
+        strm << "create socket failed: " << errnoText;
+        SCX_BOOKEND_PRINT (strm.str ().c_str ());
+    }
+#else // PRINT_BOOKENDS
+            }
+        }
+    }
+#endif // PRINT_BOOKENDS
+    return result;
+}
+
+
+class pending_client
+{
+public:
+    enum Result
+    {
+        VALID,
+        INVALID,
+        INCOMPLETE,
+    };
+
+    /*ctor*/ pending_client (int fd)
+        : m_remaining (sizeof (m_key))
+        , m_fd (fd)
+    {
+        SCX_BOOKEND ("pending_client::ctor");
+        memset (m_key, 0, sizeof (m_key));
+    }
+
+    int get_fd () { return m_fd; }
+    void close_socket ()
+    {
+        SCX_BOOKEND ("pending_client::close_socket");
+        close (m_fd);
+    }
+    int read_and_validate (unsigned int (&key)[4]);
+
+private:
+    unsigned int m_key[4];
+    ssize_t m_remaining;
+    int m_fd;
+};
+
+
+int
+pending_client::read_and_validate (
+    unsigned int (&key)[4])
+{
+    SCX_BOOKEND ("pending_client::read_and_validate");
+    int result = INCOMPLETE;
+    if (0 < m_remaining)
+    {
+        ssize_t nRead = read (
+            m_fd,
+            reinterpret_cast<unsigned char*>(
+                m_key) + (sizeof (m_key) - m_remaining),
+            m_remaining);
+        if (0 < nRead)
+        {
+            m_remaining -= nRead;
+        }
+        else if (EINTR != errno ||
+                 0 == nRead)
+        {
+            SCX_BOOKEND_PRINT ("error on read");
+            close (m_fd);
+            result = INVALID;
+        }
+    }
+    if (0 == m_remaining)
+    {
+        result = std::equal (key, key + 4, m_key) ? VALID : INVALID;
+        if (VALID == result)
+        {
+            SCX_BOOKEND_PRINT ("VALID");
+        }
+        else
+        {
+            SCX_BOOKEND_PRINT ("INVALID");
+        }
+    }
+    return result;
+}
+
+
+int
+accept_socket (
+    int listenerFD,
+    std::list<pending_client>* pPendingClients)
+{
+    SCX_BOOKEND ("accept_socket");
+    int result = EXIT_SUCCESS;
+    bool done = false;
+    do
+    {
+        sockaddr_in addr;
+        socklen_t addrlen = sizeof (addr);
+        int fd = accept (listenerFD, reinterpret_cast<sockaddr*>(&addr),
+                         &addrlen);
+        if (-1 != fd)
+        {
+            int flags = fcntl (fd, F_GETFL, 0);
+            if (-1 != flags &&
+                -1 != fcntl (fd, F_SETFL, flags | O_NONBLOCK))
+            {
+                pPendingClients->push_back (pending_client (fd));
+            }
+            else
+            {
+                // failed to set the socket to non-blocking
+                SCX_BOOKEND_PRINT ("failed to set new socket to non-blocking");
+                close (fd);
```

Comment:
break here?

Reply:
This is a case that ~should not~ ever occur.  And if it does, it is not a terminal condition.  There could be an incoming connection that fails.  This handles the failed connection and continues to receive other incoming connections.  This is as designed.